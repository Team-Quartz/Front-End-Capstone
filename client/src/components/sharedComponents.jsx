import react from 'react';
import styled from 'styled-components';

const FlexRow = styled.div`
  display: flex;
  flex-direction: row;
`;

const StarBounds = styled.div`
  width: 1em;
  height: 1em;
  overflow: hidden;
  position: relative;
`;

const StarImg = styled.img`
  position: absolute;
  top: 0;
  left: -${({ amount }) => amount}em;
  height: 1em;
  width: 5em;
`;

export const Stars = ({ reviewsMeta }) => {
  const ratingClipped = Math.floor(reviewsMeta.averageRating * 4);
  const stars = [];
  for (let i = 0; i < 20; i += 4) {
    stars.push(
      <StarBounds key={i}>
        <StarImg src='./img/stars.png' amount={Math.max(0, Math.min(4, ratingClipped - i))} />
      </StarBounds>
    );
  }
  return <FlexRow>{stars}</FlexRow>;
};


const ModalMain = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  background: rgba(0, 0, 0, 0.5);
`;

const ModalBody = styled.div`
  position: fixed;
  width: auto;
  height: auto;
  background: white;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
  padding: 2em;
`;

export function Modal({ show, onClose, children }) {
  if (show) {
    return (
      <ModalMain>
        <ModalBody>
          <button onClick={onClose}>X</button>
          {children}
        </ModalBody>
      </ModalMain>
    );
  }
  return '';
}
