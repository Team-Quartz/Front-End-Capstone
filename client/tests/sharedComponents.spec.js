/**
 * @jest-environment jsdom
 */
import React from 'react';
import reactDOM from 'react-dom';
import { Stars } from '../src/components/sharedComponents.jsx';
import renderer from 'react-test-renderer';

describe('Shared Components', () => {
  let stars;
  beforeAll(() => {
    //this setup uses jsdom to generate a headless browser instance, and render the Stars object into it
    //then we can use regular js DOM manipulation to get information about how it rendered and test on that
    stars = document.createElement('div');
    document.body.appendChild(stars);
    reactDOM.render(<Stars reviewsMeta={{ averageRating: 3.25 }} />, stars);
  });
  describe('Stars component, jsdom', () => {
    it('should have five children, nested in a div', () => {
      expect(stars.children[0].children.length).toEqual(5);
    });
    it('should have images as children', () => {
      expect(stars.children[0].children[0].children[0].tagName).toEqual('IMG');
    });
  });
  describe('Stars component, react-test-renderer', () => {
    it('renders correctly', () => {
      //this test uses react-test-renderer to generate a json object version of the DOM
      //the snapshot method compares it to a saved, known-correct version
      //but also regular object traversal can be used to check for specific properties, etc
      const tree = renderer.create(<Stars reviewsMeta={{ averageRating: 3.25 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
