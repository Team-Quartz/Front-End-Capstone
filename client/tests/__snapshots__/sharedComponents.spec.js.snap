// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shared Components Stars component, react-test-renderer renders correctly 1`] = `
<div
  className="sc-bdvvtL kOOFdf"
>
  <div
    className="sc-gsDKAQ exZWXk"
  >
    <img
      className="sc-dkPtRN sAWEY"
      src="./img/stars.png"
    />
  </div>
  <div
    className="sc-gsDKAQ exZWXk"
  >
    <img
      className="sc-dkPtRN sAWEY"
      src="./img/stars.png"
    />
  </div>
  <div
    className="sc-gsDKAQ exZWXk"
  >
    <img
      className="sc-dkPtRN sAWEY"
      src="./img/stars.png"
    />
  </div>
  <div
    className="sc-gsDKAQ exZWXk"
  >
    <img
      className="sc-dkPtRN htyiDR"
      src="./img/stars.png"
    />
  </div>
  <div
    className="sc-gsDKAQ exZWXk"
  >
    <img
      className="sc-dkPtRN jEbGhI"
      src="./img/stars.png"
    />
  </div>
</div>
`;
